Modified check_type(...) to add "but found <type>"

jjberry314 opened the pull request:
The overload for a single type adds "but found <type>"  the end of the string, but the overload for multiple types does not. Now it does.

jjberry314 commented:
My proposed changes to empty() and size() need to be reviewed by Travis in case he is relying on the previous behaviour elsewhere in the library. I realise that there is a backward compatibility risk here, so I am happy for this change to be denied. 

I think my change to check_type() is benign though.

jjberry314 commented:
I think that, on reflection, perhaps size() == 0 is not valid for a value of type null, since its size is indeterminate rather than zero, strictly speaking. 

I do still think that empty() should return true for a null type though.

jjberry314 commented:
The extra debug seemed useful to me, and also consistent with the other overload.

jjberry314 commented:
As for the version number, I agree with you. Not throwing an exception when before it did is a minor change. It's unlikely that someone would be relying on that behaviour. 
I don't think it warrants a major version number change as it is not a major change in behaviour.

tgockel commented on src/jsonv/value.cpp:
```diff
@@ -503,10 +503,11 @@ bool value::empty() const
     case jsonv::kind::boolean:
-    case jsonv::kind::null:
     default:
         // Should never hit this...
```
Just a note, your changes will make this comment inaccurate, since the `false` right below will be proper.

jjberry314 commented:
I've made the change to empty(). The return value of default: is largely irrelevant as unless you add a new kind, it will never be hit. 
I've also modified the doxygen comments for it. 

I noticed that size() is returning a boolean false implicitly cast to size_t as its default return value, and I know some compilers will issue a warning about this, so I changed it to a 0.
